sdk/eventhub/Azure.Messaging.EventHubs/perf/Scenarios/BufferedPublishToPartitionsTest.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Azure.Messaging.EventHubs.Producer;
+
+namespace Azure.Messaging.EventHubs.Perf.Scenarios
+{
+    /// <summary>
+    ///   The performance test scenario based on publishing events with the
+    ///   <see cref="EventHubBufferedProducerClient" /> using partition keys for
+    ///    partition assignment.
+    /// </summary>
+    ///
+    /// <seealso cref="BufferedPublishPerfTest{TOptions}" />
+    ///
+    public class BufferedPublishToPartitionsTest : BufferedPublishPerfTest<EventHubsBufferedOptions>
+    {
+        /// <summary>The random number generator to use for this test.</summary>
+        private Random _random = new();
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="BufferedPublishToPartitionsTest"/> class.
+        /// </summary>
+        ///
+        /// <param name="options">The set of options to consider for configuring the scenario.</param>
+        ///
+        public BufferedPublishToPartitionsTest(EventHubsBufferedOptions options) : base(options)
+        {
+        }
+
+        /// <summary>
+        ///   Determines the set of <see cref="EnqueueEventOptions"/> needed for this test scenario.
+        /// </summary>
+        ///
+        /// <returns>The set of options to use when enqueuing events to be published.</returns>
+        ///
+        /// <remarks>
+        ///     This method is expected to be called each time that a set of events
+        ///     is to be enqueued; it should be kept light and fast.
+        /// </remarks>
+        ///
+        protected override EnqueueEventOptions CreateEnqueueOptions() =>
+            new EnqueueEventOptions
+            {
+                PartitionId = Partitions[_random.Next(0, Partitions.Length - 1)]
```

Comment:
The existing perf test `PublishEventsToPartition` uses `ParallelIndex` to determine the partition, so each parallel instance will always publish to the same partition, and you need to use N parallel instances to publish to N partitions:

https://github.com/Azure/azure-sdk-for-net/blob/c56d65f9357b7c5f64aa7ed2fa31f0098480ee74/sdk/eventhub/Azure.Messaging.EventHubs/perf/Scenarios/PublishEventsToPartition.cs#L37-L43

I don't have a preference which is better, but I think both tests should use the same algorithm, so we can get a fair comparison between the two publishers.

Reply:
Yeah, there's some differences between the types that we may or may not need to account for.   I'm going to drop 30 on your calendar next week;  I'd like to run the entire context by you for your thoughts.